taroscript/send.go

```diff
@@ -290,154 +164,181 @@ func IsValidInput(input *commitment.TaroCommitment,
-			[][32]byte{senderStateKey, receiverStateKey},
-		)
+	// This should be caught way earlier but just to make sure that we never
+	// overflow when converting the input amount to int64 we check this
+	// again.
+	inputAsset := input.Asset()
+	if inputAsset.Amount > math.MaxInt64 {
+		return fmt.Errorf("amount int64 overflow")
 	}
 
-	senderLocator := updatedDelta.Locators[senderStateKey]
-	receiverLocator := updatedDelta.Locators[receiverStateKey]
+	// A collectible cannot be split into individual pieces. So there can
+	// only be a tombstone and a recipient output.
+	if inputAsset.Type == asset.Collectible && len(outputs) > 2 {
+		return ErrInvalidCollectibleSplit
+	}
 
-	inputAsset := updatedDelta.InputAssets[prevInput]
+	var (
+		residualAmount = inputAsset.Amount
+		splitLocators  = make([]*commitment.SplitLocator, len(outputs))
+	)
+	for idx := range outputs {
+		vOut := outputs[idx]
+
+		// We assume the first output is the change output (or
+		// tombstone if there is no change in a non-interactive send).
+		if idx == 0 {
+			// The change output should always be at index 0.
+			if !vOut.IsChange {
+				return ErrInvalidChangeOutputLocation
+			}
 
-	// Populate the remaining fields in the splitLocators before generating
-	// the splitCommitment.
-	senderLocator.AssetID = addr.ID()
-	senderLocator.ScriptKey = asset.ToSerialized(&scriptKey)
-	senderLocator.Amount = inputAsset.Amount - addr.Amount
-	updatedDelta.Locators[senderStateKey] = senderLocator
+			// A zero-amount change output (tombstone) must spend to
+			// the un-spendable NUMS script key.
+			if vOut.Amount == 0 &&
+				!vOut.ScriptKey.PubKey.IsEqual(asset.NUMSPubKey) {
 
-	receiverLocator.AssetID = addr.ID()
-	receiverLocator.ScriptKey = asset.ToSerialized(&addr.ScriptKey)
-	receiverLocator.Amount = addr.Amount
-	updatedDelta.Locators[receiverStateKey] = receiverLocator
+				return commitment.ErrInvalidScriptKey
+			}
+		} else {
+			if vOut.IsChange {
+				return ErrInvalidChangeOutputLocation
+			}
+		}
 
-	// Enforce an unspendable root split if the split sends the full value
-	// of the input asset or if the split sends a collectible.
-	if (senderLocator.Amount == 0 || inputAsset.Type == asset.Collectible) &&
-		senderLocator.ScriptKey != asset.NUMSCompressedKey {
+		residualAmount -= vOut.Amount
+
+		locator := vOut.SplitLocator(inputAsset.ID())
+		splitLocators[idx] = &locator
+	}
+
+	// We should now have exactly zero value left over after splitting.
+	if residualAmount != 0 {
+		return ErrInvalidSplitAmounts
+	}
+
+	// If we have an interactive full value send, we don't need a tomb stone
+	// at all.
+	inputIDCopy := input.PrevID
+	if interactiveFullValueSend(input, outputs) {
+		// We'll now create a new copy of the old asset, swapping out
+		// the script key. We blank out the tweaked key information as
+		// this is now an external asset.
+		outputs[1].Asset = inputAsset.Copy()
+		outputs[1].Asset.ScriptKey = outputs[1].ScriptKey
```

Comment:
IIUC, that's done as this is the top level `ScriptKey` struct that contains the raw key and tweaked key (tho in this case only the tweaked key matters).

Reply:
Yeah, on the output level we only really need the full output key. But if we do have the derivation information (because the key is local to our node), we want to serialize it in the PSBT. But it's optional, really.

Reply:
Ok, so we set `outputs[0].ScriptKey` earlier when creating the VPacket, and are now moving it to the receiver asset? And that key likely has no tweak information.

Reply:
Yes, correct. Because at the point where you create the virtual packet, there is no asset to set the script key on yet. That's why it's a separate field to store it until we've created the output asset.